### ISSUE TITLE
Glow 1 - Basic Steps

### ISSUE THREAD

**Issue by ytgui** [bosskwei/temp#29]
# Glow
+ A **machine learning compiler** and **execution engine** for various hardware targets.
+ A **backend** for high-level machine learning frameworks.
+ State of the art **compiler optimizations** and **code generation** of neural network graphs.

## How
+ Lowers a traditional neural network dataflow graph into a two-phase strongly-typed IR.
+ High-level IR -> domain-specific optimizations (maybe for special network structure and operation).
+ Lower-level instruction-based address-only IR -> memory-related optimizations (instruction scheduling, static memory allocation and copy elimination).
+ Lowest level -> machine-specific code generation (take advantage of specialized hardware features).

![](https://github.com/pytorch/glow/raw/master/docs/3LevelIR.png)

## Feature
+ Support a high number of input operators.
+ Support  a large number of hardware targets.
+ Eliminating the need to implement all operators on all targets.
+ Reduce the input space to focus on a small number of linear algebra primitives.

**Comment by ytgui** [comment]
## Demo
1. Init backend config as `llvm::cl::opt`
```CPP
namespace {
llvm::cl::OptionCategory mnistCat("MNIST Options");
llvm::cl::opt<BackendKind> executionBackend(
    llvm::cl::desc("Backend to use:"),
    llvm::cl::Optional,
    llvm::cl::values(
        clEnumValN(BackendKind::Interpreter, "interpreter", "Use interpreter"),
        clEnumValN(BackendKind::CPU, "cpu", "Use CPU"),
        clEnumValN(BackendKind::OpenCL, "opencl", "Use OpenCL")),
    llvm::cl::init(BackendKind::CPU),
    llvm::cl::cat(mnistCat));
}
```

2. Context and Execution
```CPP
Context ctx;
ExecutionEngine EE(executionBackend);
```

3. Build graph
```CPP
// High level program and graph
auto &mod = EE.getModule();
Function *F = mod.createFunction("main");

// Construct the network
Placeholder *A = mod.createPlaceholder(ElemKind::FloatTy, {minibatchSize, 28, 28, 1}, "input", false);
auto *CV0 = F->createConv(ctx, "conv", A, 16, 5, 1, 2, 1);
auto *RL0 = F->createRELU("relu", CV0);
auto *MP0 = F->createMaxPool("pool", RL0, 3, 3, 0);

auto *CV1 = F->createConv(ctx, "conv", MP0, 16, 5, 1, 2, 1);
auto *RL1 = F->createRELU("relu", CV1);
auto *MP1 = F->createMaxPool("pool", RL1, 3, 3, 0);

auto *FCL1 = F->createFullyConnected(ctx, "fc", MP1, 10);
Placeholder *selected = mod.createPlaceholder(ElemKind::Int64ITy, {minibatchSize, 1}, "selected", false);
auto *SM = F->createSoftMax("sm", FCL1, selected);
SaveNode *result = F->createSave("return", SM);

// The training configuration
TrainingConfig TC;
TC.learningRate = 0.001;
TC.momentum = 0.9;
TC.L2Decay = 0.001;
TC.batchSize = minibatchSize;

// Autograd
Function *TF = glow::differentiate(F, TC);
EE.compile(CompilationMode::Train, TF);
```

4. Feed data
```CPP
Tensor imageInputs;
Tensor labelInputs;
unsigned numImages = loadMNIST(imageInputs, labelInputs);

// Allocation
Tensor *inputTensor = ctx.allocate(A);
Tensor *resultTensor = ctx.allocate(result->getPlaceholder());
ctx.allocate(selected);
```

5. Training
```CPP
for (int epoch = 0; epoch < 60; epoch++) {
  // On each training iteration take a slice of imageInputs and labelInputs
  // and put them into variables A and B, then run forward and backward passes
  // and update weights.
  runBatch(EE, ctx, numIterations, sampleCounter, {A, selected}, {&imageInputs, &labelInputs});
}
```

6. Inference
```CPP
glow::convertPlaceholdersToConstants(F, ctx, {A, result->getPlaceholder()});
EE.compile(CompilationMode::Infer, F);

auto LIH = labelInputs.getHandle<int64_t>();

int rightAnswer = 0;

for (int iter = numIterations; iter < numIterations + 10; iter++) {
  inputTensor->copyConsecutiveSlices(&imageInputs, minibatchSize * iter);
  EE.run(ctx);

  for (unsigned i = 0; i < minibatchSize; i++) {
    auto T = resultTensor->getHandle<>().extractSlice(i);
    int64_t guess = T.getHandle<>().minMaxArg().second;

    int64_t correct = LIH.at({minibatchSize * iter + i, 0});
    rightAnswer += (guess == correct);
  }
}
```

**Comment by bosskwei** [comment]
## ExecutionEngine
## Placeholder
## Function
## SaveNode

**Comment by ytgui** [comment]
## CPU
```bash
===-------------------------------------------------------------------------===
                         Miscellaneous Ungrouped Timers
===-------------------------------------------------------------------------===
   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
  10.5689 (100.0%)   0.0001 (100.0%)  10.5690 (100.0%)  10.5692 (100.0%)  Training
  10.5689 (100.0%)   0.0001 (100.0%)  10.5690 (100.0%)  10.5692 (100.0%)  Total
```

## OpenCL
```bash
===-------------------------------------------------------------------------===
                         Miscellaneous Ungrouped Timers
===-------------------------------------------------------------------------===
   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
  10.8819 (100.0%)   2.8823 (100.0%)  13.7642 (100.0%)  13.4875 (100.0%)  Training
  10.8819 (100.0%)   2.8823 (100.0%)  13.7642 (100.0%)  13.4875 (100.0%)  Total
```
